rust-runtime/aws-smithy-http-server/src/main.rs

```diff
@@ -0,0 +1,58 @@
+use std::sync::{Arc, Mutex};
+
+use aws_smithy_http_server::add_extension::AddExtensionLayer;
+use aws_smithy_http_server::extension::Extension;
+use aws_smithy_http_server::model::*;
+use aws_smithy_http_server::operation_registry::SimpleServiceOperationRegistryBuilder;
+use aws_smithy_http_server::routing::Router;
+
+struct State {
+    counter: Mutex<i32>,
+}
+
+async fn _healthcheck_operation(_input: HealthcheckInput) -> HealthcheckOutput {
+    HealthcheckOutput
+}
+
+async fn healthcheck_operation_with_state(_input: HealthcheckInput, state: Extension<Arc<State>>) -> HealthcheckOutput {
+    let mut counter = state.0.counter.lock().unwrap();
+    *counter += 1;
+    dbg!(&counter);
+
+    HealthcheckOutput
+}
+
+async fn register_service_operation(
+    _input: RegisterServiceInput,
+) -> Result<RegisterServiceOutput, RegisterServiceError> {
+    Ok(RegisterServiceOutput)
+}
+
+async fn _register_service_operation_with_state(
+    _input: RegisterServiceInput,
+    _state: Extension<State>,
+) -> Result<RegisterServiceOutput, RegisterServiceError> {
+    Ok(RegisterServiceOutput)
+}
+
+#[tokio::main]
+async fn main() {
+    let app: Router = SimpleServiceOperationRegistryBuilder::default()
+        // .health_check(_healthcheck_operation)
```

Comment:
You can swap these lines out and the application runs fine, independently of whether you provide a handler requiring state or not.

Reply:
Nice work!